csharp/src/Ice/IncomingFrame.cs

```diff
@@ -36,19 +38,23 @@ public abstract class IncomingFrame
         private readonly int _maxSize;
 
         /// <summary>Decompresses the encapsulation payload if it is compressed. Compressed encapsulations are only
-        /// supported with 2.0 encoding.</summary>
+        /// supported with the 2.0 encoding.</summary>
         public void DecompressPayload()
         {
-            if (!HasCompressedPayload)
+            if (PayloadCompressionFormat == CompressionFormat.Decompressed)
             {
                 throw new InvalidOperationException("the encapsulation's payload is not compressed");
             }
+            else if (PayloadCompressionFormat != CompressionFormat.GZip)
+            {
+                throw new NotSupportedException($"cannot decompress compression format `{PayloadCompressionFormat}'");
+            }
             else
             {
                 int encapsulationOffset = this is IncomingResponseFrame ? 1 : 0;
 
                 ReadOnlySpan<byte> buffer = Payload.Slice(encapsulationOffset);
-                (int size, int sizeLength, Encoding _) = buffer.ReadEncapsulationHeader(Protocol.GetEncoding());
+                (int _, int sizeLength, Encoding _) = buffer.ReadEncapsulationHeader(Protocol.GetEncoding());
```

Comment:
```suggestion
                int sizeLength = buffer.ReadEncapsulationHeader(Protocol.GetEncoding()).SizeLenght;
```

Reply:
Even simpler:
```
int sizeLength = Protocol == Protocol.Ice1 ? 4 : buffer[0].ReadSizeLength20();
```

Reply:
Do we need to check the protocol here, doesn't the GZip payload implies Ice2 protocol?

Reply:
This compression is about the encoding of the payload.

For example, we could have an ice1 request with 2.0-encoded args that are compressed. This would typically occur in a protocol-bridging situation, where we forward a payload mostly as-is across different versions of the Ice protocol. During this bridging, we only convert the encapsulation header (size + encoding) from ice1's header encoding (1.1) to ice'2 header encoding (2.0) or vice-versa. The remaining bytes (such as compression format, decompressed size and compressed bytes) don't change.